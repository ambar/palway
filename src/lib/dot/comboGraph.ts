import getPalIcon from '../getPalIcon'
import type {PalName} from '../palNames'
import {normalPalsByName} from '../pals'
import {styleMap, styleToAttr} from './shared'
import type {RankDir} from './dot'

type Options = {
  preset?: 'noImage' | 'image'
  direction?: RankDir
}

/**
 * Create a combo graph of all possible matches for a given child
 * @see https://graphviz.org/docs/graph/
 */
export const makeComboGraph = (
  parents: [PalName, PalName][],
  child: PalName,
  i18n: (key: string) => string,
  {preset = 'image', direction = 'LR'}: Options = {},
) => {
  const style = styleMap[preset]
  const attr = (palName: PalName) => {
    return [
      preset === 'image' &&
        `image="${getPalIcon(normalPalsByName[palName].pal_dev_name)}"`,
      `label="${name(palName)}"`,
    ]
      .filter(Boolean)
      .join(',')
  }
  const name = (palName: PalName) => {
    const pal = normalPalsByName[palName]
    return i18n(`palName_${pal.pal_dev_name}`)
  }
  if (!parents?.length) {
    return ''
  }
  // label="Palway"
  // labelloc=r
  // labeljust=r
  // fontsize=12
  // fontname="Helvetica,Arial,sans-serif"
  return String.raw`
digraph Palway {
  node [fontname="Helvetica,Arial,sans-serif",style=dashed]
  edge [fontname="Helvetica,Arial,sans-serif",style=dashed,arrowsize=.9,arrowhead=empty]
  rankdir=${direction}
  node [${styleToAttr(style)}];
  root [${attr(child)},class="root"]

  ${parents
    .map((parent, i) => {
      const na = parent[0].replace(/ /g, '_')
      const nb = parent[1].replace(/ /g, '_')
      const connect = `${na}_${nb}_${i}`
      const a = `${na}_node_a_${i}`
      const b = `${nb}_node_b_${i}`
      return String.raw`
  ${connect}[shape=circle,style=filled,label="",width=.05, height=.05]
  ${connect} -> root []
  ${a} -> ${connect} [dir=none]
  ${b} -> ${connect} [dir=none]
  ${a} [${attr(parent[0])}]
  ${b} [${attr(parent[1])}]
`
    })
    .join('\n')}
}
`
}
