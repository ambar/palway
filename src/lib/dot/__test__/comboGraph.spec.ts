import {test, expect} from 'vitest'
import {makeComboGraph} from '../comboGraph'

test('makeComboGraph', () => {
  expect(
    makeComboGraph([['Cattiva', 'Lifmunk']], 'Vixy', v => v),
  ).toMatchSnapshot()
})
