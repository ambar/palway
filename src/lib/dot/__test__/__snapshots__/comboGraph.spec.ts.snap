// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`makeComboGraph 1`] = `
"
digraph Palway {
  node [fontname="Helvetica,Arial,sans-serif",style=dashed]
  edge [fontname="Helvetica,Arial,sans-serif",style=dashed,arrowsize=.9,arrowhead=empty]
  rankdir=LR
  node [shape=egg,height=1.5,width=1.5,labelloc=b];
  root [image="/src/assets/images/pal_icons_200w/T_CuteFox_icon_normal.png",label="palName_CuteFox",class="root"]

  
  Cattiva_Lifmunk_0[shape=circle,style=filled,label="",width=.05, height=.05]
  Cattiva_Lifmunk_0 -> root []
  Cattiva_node_a_0 -> Cattiva_Lifmunk_0 [dir=none]
  Lifmunk_node_b_0 -> Cattiva_Lifmunk_0 [dir=none]
  Cattiva_node_a_0 [image="/src/assets/images/pal_icons_200w/T_PinkCat_icon_normal.png",label="palName_PinkCat"]
  Lifmunk_node_b_0 [image="/src/assets/images/pal_icons_200w/T_Carbunclo_icon_normal.png",label="palName_Carbunclo"]

}
"
`;
